File: python/lcskpp.py
from fenwick_tree import Fenwick_tree
from match_pairs import MatchPairs

def LCSkpp(k,string_1,string_2):
    #LCSkpp algorith implementation
    n=len(string_2)
    m=len(string_1)
    maxColDP=Fenwick_tree(n)
    #maxColDP is a Fenwick tree for fas calculation of maxRange query in the range of [0,j]
    events,r=MatchPairs(string_1,string_2,k)
    #events is a sorted list of all the substring pair starts and ends in row major ordering
    dp={}
    #dp is a map of sunsequence values per event start len(dp)=r(number of Match Pairs)
    continue_map={}
    #continue map every start that exists is stored on the (i+1,j+1) place so if (i+1,j+1)
    # exists we can get the sequence it continues in O(log n) time

    for event in events:
        if event[2]==True: #if event is start
            value,parent=maxColDP.get(event[1])
            dp[event]=(k+value,parent) #O(log n) time
            continue_map[(event[0]+1,event[1]+1,event[2])]=event #create a map for every seen event
            #if the event is seen we create a i+1,j+1 bucket for it if i+1,j+1 exists we can find
            # its continuation in O(1) time
        else: #if event is end
            p=(event[0]-k,event[1]-k,True) #find substring start
            g=continue_map.get(p) #find the sequence it continues
            if g!=None:
                #update if continues and is greater than previous
                if dp[g][0]+1>dp[p][0]:
                    dp[p]=(dp[g][0]+1,g)
            #update the Fenwick tree with the longest match
            maxColDP.update(event[1],dp[p][0],p)

    #path reconstruction
    maximum_length=0
    child_first=(0,0,True)
    keys=[key for key in dp.keys()]
    keys.sort() #we sort the keys to find the biggest (row_major) index that has the maxValue
    for child in keys:
        value,p=dp[child]
        if value>=maximum_length:
            maximum_length=value
            child_first=child
    path=[(child_first[0],child_first[1])] #for the max value return the parent node
    child=child_first

    parent=(0,0,True)
    if maximum_length==0:
        return

    while(True):
        #walk through the parent child chain and remember the path
        v,parent=dp[child]
        if (parent==(-1,-1,True)):
            break
        #path.append((parent[0],parent[1]))
        path.append((parent[0],parent[1]))
        if (parent==(0,0,True)):
            break
        child=parent
    #reverse the path /we can also sort
    path.reverse()
    i_prev=j_prev=-k
    s1_m=s2_m=""
    test_s1=test_s2=""
    for step in path:
        i,j=step
        cri=chi=crj=chj=0
        #this is for - that indicate which characters we take from the sequence only for tests
        #not part of the final return
        if i-i_prev>=k:
            cri=i-i_prev-k
        else:
            chi=k-1
        if j-j_prev>=k:
            crj=j-j_prev-k
        else:
            chj=k-1

        s1_m+="-"*cri
        s2_m+="-"*crj
        #tests are the true sequences
        test_s1+=string_1[i+chi:i+k]
        test_s2+=string_2[j+chj:j+k]

        s1_m+=string_1[i+chi:i+k]
        s2_m+=string_2[j+chj:j+k]
        i_prev,j_prev=i,j

    s1_m+="-"*(m-len(s1_m))
    s2_m+="-"*(n-len(s2_m))
    f=open("py_output.txt","w")
    #f.write(string_1+"\n")
    #f.write(s1_m+"\n\n")

    #f.write(string_2+"\n")
    #f.write(s2_m+"\n\n")

    f.write(test_s1+"\n\n")
    #f.write(str(len(test_s1))+"\n\n")

    f.write(str(maximum_length)+"\n")

    return

File: java/LCSkpp.java
/** @author Toni Bakarcic */
/** class for implementing the LCSk++ algorithm*/

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class LCSkpp {
	
	// private variable that store the longest found match
	private int maxLen;
	// other helper private variables
	private Pair first;
	private String s1m = "", s2m = "", test1m = "", test2m = "";
	private int iPrev, jPrev;

	// the constructor for running the algorithm
	public LCSkpp(int k, String x, String y) {
		int n = y.length();
		// a FenwickTree variable that is used for fast calculation of the maxRange query in range [0,j]
		FenwickTree maxColDP = new FenwickTree(n);
		MatchPairs mp = new MatchPairs(k, x, y);
		// An ArrayList of all matched substring Pairs 
		ArrayList<Pair> events = mp.getPairs();
		// A HashMap for storing all the Pairs along with their lengths
		HashMap<Pair, Node> dp = new HashMap<Pair, Node>();
		// A HashMap for storing all specific Pairs that continue some other Pair
		HashMap<Pair, Pair> continueMap = new HashMap<Pair, Pair>();
		// An ArrayList that has all the substrings inside the input file that create the longest matching set
		ArrayList<Pair> path = new ArrayList<Pair>();

		for (int i = 0; i < events.size(); i++) {
			Pair event = events.get(i);
			// check if the event is the start
			if (event.getBool() == true) {
				// complexity is O(log n)
				Node temp = maxColDP.get(event.getJ());
				continueMap.put(new Pair(event.getI() + 1,  event.getJ() + 1, event.getBool()), event);
				dp.put(event, new Node(temp.getLen() + k, temp.getI(), temp.getJ(), temp.getBool()));
			}else {
				// if the event is not at the start, go to the start
				Pair p = new Pair(event.getI() - k, event.getJ() - k, true);
				// complexity is O(log n)
				Object p2 = continueMap.get(p);
				if(p2 != null) {
					// update the value if it continues
					int a=dp.get((Pair)p2).getLen() + 1;
					int b=dp.get(p).getLen();
					if (a>b){
						dp.put(p, new Node(a, (Pair)p2));
					}
				}
				maxColDP.update(event.getJ(), dp.get(p).getLen(), p);
			}
		}

		// reconstruction of the "Path"
		this.maxLen = 0;
		first = new Pair(0, 0, true);
		ArrayList<Pair> keys = new ArrayList<Pair>();
		keys.addAll(dp.keySet());
		Collections.sort(keys);
		keys.forEach((Pair child) -> {
			Node temp = dp.get(child);
			if (temp.getLen() >= this.maxLen) {
				this.maxLen = temp.getLen();
				first = child;
			}
		});
		path.add(first);
		Pair child = this.first;

		Pair parent = new Pair(0, 0, true);
		while(true) {
			Node temp = dp.get(child);
			if (temp != null)
				parent = new Pair(temp.getI(), temp.getJ(), temp.getBool());
			if ((parent.getI() == -1 && parent.getJ() == -1 && parent.getBool() == true)
					|| (parent.getI() == 0 && parent.getJ() == 0 && parent.getBool() == true))
				break;
			path.add(parent);
			child = parent;
		}
		Collections.sort(path);
		this.iPrev = -k;
		this.jPrev = -k;
		path.forEach((Pair step) -> {
			int i = step.getI();
			int j = step.getJ();
			int cri = 0, chi = 0, crj = 0, chj = 0;
			if (i - iPrev >= k) {
				cri = i - iPrev - k;
			}else {
				chi = k - 1;
			}
			if (j - jPrev >= k) {
				crj = j - jPrev - k;
			}else {
				chj = k - 1;
			}
			this.s1m += generateString("-", cri);
			this.s2m += generateString("-", crj);
			this.test1m += x.substring(i + chi, i + k);
			this.test2m += y.substring(j + chj, j + k);
			this.s1m += x.substring(i + chi, i +k);
			this.s2m += y.substring(j + chj, j +k);
			this.iPrev = i;
			this.jPrev = j;
		});

		this.s1m += generateString("-", x.length() - s1m.length());
		this.s2m += generateString("-", y.length() - s2m.length());

		// Write the results of the algorithm to the output file
		try {
			File file = new File("output.txt");

			FileWriter fw = new FileWriter(file);
			fw.write(s1m.length());
			fw.write("\n");
			fw.write(x);
			fw.write(s1m);
			fw.write("\n");
			fw.write(y);
			fw.write(s2m);
			fw.write("\n");
			fw.write(test1m.length());
			fw.write(test1m);
			fw.write(test2m);
			fw.write(String.valueOf(test1m.equals(test2m)));
			fw.write("\n");
			fw.close();
		} catch (IOException e) {
			System.out.println("Output file error.");
		}
	}

	public String generateString(String s, int n) {
		String str = "";
		for (int i=0; i<n; i++)
			str += s;
		return str;
	}

	public int getLCSkpp () {
		return this.maxLen;
	}

}
